torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -155,6 +155,139 @@ std::pair<std::string, std::string> codeGeneration(Fusion& fusion) {
   return std::make_pair(func_name, str_stream.str());
 };
 
+bool validateKernelArgTensor(
+    const at::Tensor& arg,
+    const Val* param,
+    int device_index,
+    std::stringstream& msg) {
+  // Arg is a tensor. Param must be a tensor too.
+  if (*param->getValType() != ValType::TensorView) {
+    msg << "Argument is a tensor, but the parameter is not.";
+    return false;
+  }
+  // Check the rank of the tensors.
+  size_t arg_dim = arg.dim();
+  const auto& root_dom = *static_cast<const TensorView*>(param)->getRootDomain();
```

Comment:
I'm not quite sure about this.

Prior to reduction, I think this is all right. I used the same snippet to set rank on IO tensors.
Right now I'm worried that `getRootDomain` could be affected by reduction later and have dimension collapsed.

Per the new code here: https://github.com/csarofeen/pytorch/blob/2e8f8628bb8f2384b064f116c16414baa7c6c6c8/torch/csrc/jit/codegen/cuda/ir_iostream.cpp#L39-L43

I think we should change this line to 
`const auto& root_dom = *static_cast<const TensorView*>(param)->getRootDomain().noReductions();`
cc'ing @csarofeen in case I'm talking non-sense

Reply:
Looking at the way you are excluding the reduction from few lines below, I think this is doing the same thing as `noReductions()`. But could be cleaner using that instead.

Reply:
But I don't quite understand why reduction dimensions are collapsed for IO tensor. That might be a topic better discussed in slack I'll start a thread for that.

Reply:
NVM, I think now I get the use case here.
Please ignore the previous message except the first one. You might still want to opt for `noReductions()` for shorter code.

Reply:
Thanks. Didn't know `noReductions`. Fixed.